### PR TITLE
add report only csp policy and report endpoint

### PR DESCRIPTION
**Pull request by deathbearbrown**
Nightmare linter strikes again.

Loves them 4 space tabs. 

Common.inc line 62-74 is the csp header and I added a report endpoint which will create a log file.

### CODE REVIEW

File: www/csp-violation-report-endpoint.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+// We pretty print the JSON before adding it to the log file
+if ($json_data = json_decode($json_data)) {
+    $json_data = json_encode($json_data, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES);
+
+    if (!file_exists($log_file)) {
+        // Send an email
+        $message = "The following Content-Security-Policy violation occurred on " .
+            $current_domain . ":\n\n" .
+            $json_data .
+            "\n\nFurther CPS violations will be logged to the following log file, but no further email notifications will be sent until this log file is deleted:\n\n" .
```
Comment (tkadlec) [inline www/csp-violation-report-endpoint.php:26]: nit, but "Further CSP violations" :)

File: www/common.inc [hunk 2]
```diff
@@ -58,8 +58,25 @@ $request_context = new RequestContext($_REQUEST, $_SERVER);
 // Disable caching by default
 header("Cache-Control: no-store, no-cache, must-revalidate, max-age=0", true);
 
+// add CSP report only
+header("Content-Security-Policy-Report-Only: default-src 'self'; 
+script-src 'report-sample' 'self' https://g.3gl.net/jp/8627/v3.3.4/M https://js.braintreegateway.com/web/dropin/1.33.0/js/dropin.min.js https://ssl.google-analytics.com/ga.js;
```
Comment (tkadlec) [inline www/common.inc:62]: I'm wondering if we should include `unsafe-inline` for both `script-src` and `style-src` for now. If we don't this is going to get _very_ noisy right away. We can focus on fixing up the other stuff first and circle back to these after the rest of the policy is mostly set.

Maybe even a separate report-only policy at that point.

Comment (tkadlec) [reply www/common.inc:62]: Also, for `script-src` specifically, probably good enough to just use the origin and ditch the full path?

File: www/csp-violation-report-endpoint.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
+<?php
+
+// Start configure
+$log_file = dirname(__FILE__) . '/csp-violations.log';
```
Comment (jefflembeck) [inline www/csp-violation-report-endpoint.php:4]: Let's make sure we add this to the .gitignore so we don't have any issues with our deployments

File: www/csp-violation-report-endpoint.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    $json_data = json_encode($json_data, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES);
+
+    if (!file_exists($log_file)) {
+        // Send an email
+        $message = "The following Content-Security-Policy violation occurred on " .
+            $current_domain . ":\n\n" .
+            $json_data .
+            "\n\nFurther CPS violations will be logged to the following log file, but no further email notifications will be sent until this log file is deleted:\n\n" .
+            $log_file;
+        mail(
```
Comment (jefflembeck) [inline www/csp-violation-report-endpoint.php:28]: Does this use the main (blocking) process? I'm worried about having 2000 users on the site and all of these going off at the same time.

File: www/csp-violation-report-endpoint.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
+<?php
+
+// Start configure
+$log_file = dirname(__FILE__) . '/csp-violations.log';
+$log_file_size_limit = 1000000; // bytes - once exceeded no further entries are added
+$email_address = 'slockwood@catchpoint.com';
```
Comment (jefflembeck) [inline www/csp-violation-report-endpoint.php:6]: oh god sue i'm so sorry

Comment (deathbearbrown) [reply www/csp-violation-report-endpoint.php:6]: haha yeah I was going to ask.. can we use a different email

File: www/csp-violation-report-endpoint.php (hunk 3 above)
Comment (jefflembeck) [reply www/csp-violation-report-endpoint.php:4]: Yeah for sure

**Comment by deathbearbrown** [comment]
Setting this to BLOCKED, because it requires some server configurations for logging.

**Pull request closed without merging**